lib/AST/Decl.cpp

```diff
@@ -1197,10 +1197,26 @@ bool PatternBindingDecl::isDefaultInitializable(unsigned i) const {
 
   // If the pattern is typed as optional (or tuples thereof), it is
   // default initializable.
-  if (auto typedPattern = dyn_cast<TypedPattern>(entry.getPattern()))
-    if (auto typeRepr = typedPattern->getTypeLoc().getTypeRepr())
+  if (auto typedPattern = dyn_cast<TypedPattern>(entry.getPattern())) {
+    if (auto typeRepr = typedPattern->getTypeLoc().getTypeRepr()) {
       if (::isDefaultInitializable(typeRepr))
         return true;
+    } else if (typedPattern->isImplicit()) {
+      // Lazy vars have implicit storage assigned to back them. Because the
+      // storage is implicit, the pattern is typed and has a TypeLoc, but not a
+      // TypeRepr.
+      //
+      // All lazy storage is implicitly default initializable, though, because
+      // lazy backing storage is optional.
+      if (auto *varDecl = typedPattern->getSingleVar())
+        // Lazy storage is never user accessible.
+        if (!varDecl->isUserAccessible()) {
+          auto type = typedPattern->getTypeLoc().getType()->getCanonicalType();
```

Comment:
No need for the `getCanonicalType` here; `getAnyOptionalObjectType` will already do it.

(In general that's true of nearly all our type operations; the ones where you actually need to get the canonical type manually are the exception rather than the rule.)

Reply:
Ah, makes sense; I'll remove that. These changes still look reasonable to you otherwise, yes?